user/pam_krb5/src/afs5log.1
.TH afs5log 8 2007/12/18 "Red Hat Linux" "System Administrator's Manual"

.SH NAME
afs5log \- AFS token initializer

.SH SYNOPSIS
afs5log [ [-v] [-5] [-k] [-p \fIpath\fP] [cell[=principal] ] ] [...]

.SH DESCRIPTION
The \fIafs5log\fP command uses Kerberos to obtain AFS tokens for the named
cells.  If no cell is named, tokens are obtained for the local cell.  If the
principal name of the \fIafs\fP service for the cell is known, the cell
name may also be given in the form \fIcell\fP=\fIprincipal_name\fP.
Attempts are made to obtain Kerberos IV credentials first, and if that fails,
\fIafs5log\fP will create Kerberos 5 (\fI2b-style\fP) tokens.

.SH ARGUMENTS
.TP
-v
Turns on verbose mode.  \fIafs5log\fP will print debugging messages while it
does its work.  Use more than once to get more detail.
.TP
-p \fIpath\fP
Determine which cell the specified \fIpath\fP resides in, and obtain tokens for
that cell.
.TP
-s \fIstrategy\fP
Overrides the default token-setting strategy
("\fIv4,524,2b,rxk5\fP") with
the named strategy.  Recognized strategy names include:
 \fIrxk5\fP  rxk5 (requires OpenAFS 1.6 or later)
 \fI2b\fP    rxkad "2b" (requires OpenAFS 1.2.8 or later)
.\"  \fI524\fP   Kerberos 524 service + traditional Kerberos IV
.\"  \fIv4\fP    traditional Kerberos IV
.TP
-5
Skip attempts to use Kerberos IV and just use \fI2b\fP-style tokens, which are
only supported by OpenAFS 1.2.8 and later.
.TP
-k
Skip attempts to use Kerberos IV and use \fIrxk5\fP-style tokens, which are
only supported by OpenAFS 1.6 and later.
.TP
-n
Assume that, if a cell name and a realm name differ only by case, that the
name of the AFS service's principal name doesn't include an instance.

.SH "SEE ALSO"
.BR pam_krb5 (5)
.BR krb5.conf (5)
.br

.SH BUGS
Probably, but let's hope not.  If you find any, please file them in the
bug database at http://bugzilla.redhat.com/ against the "pam_krb5" component.

.SH AUTHOR
Nalin Dahyabhai <nalin@redhat.com>


freeswan/CHANGES
New and notable in 1.97:

Sanity check added for at least one of AH or ESP more for checking if
"make *config" has even been run.

Fixed 2.2 local IKE fragmentation blackhole.  Still won't work if
iptraf or another pcap app is running.

MCR added ikeping(8).

Applied DHR's tunnel patch to streamline IKE/specialSA processing.

Added local and remote IKE bypass tests to the UML test suite.

UML KLIPS %trap* tests now check for upbound pf_key messages.

An ICMP ICMP_DEST_UNREACH, ICMP_PKT_FILTERED is now sent upon %reject.

Added a UML input file with first and last packets being outside the 
covered range to "sandwich" the packets under test.

The default _updown script has been cleaned up slightly, reducing
internal duplication that arose accidentally.

When doing --add on an RSA-signature conn, ipsec_auto now tells Pluto
about the keys before, not after, telling it about the conn.  The main
consequence of this is that if something is wrong with one of the keys,
the conn itself does *not* get loaded.  This should reduce confusion.

The rp_filter checks during KLIPS startup have been fixed to be a bit less
paranoid; rp_filter is thought to cause problems only on the physical
interfaces, not on the ipsecN interfaces. 

The configuration process has been fixed to be a bit more paranoid.  Some
users have had mysterious compile failures with complaints involving names
like MD5_CTX; they seem to have resulted from kernel configuration changes
somehow getting into .config but not into include/linux/autoconf.h (which
is what actually controls most of the kernel compile).  The post-config
checking now verifies not only that .config has the right stuff in it, but
that autoconf.h does too.  (We still don't see *how* this discrepancy can
arise, but this will at least catch it when it does.)

The details of linking the KLIPS source into the kernel are now under
control of a configuration variable in Makefile.inc, to make it easier
for distribution builders to change.

The top-level Makefile now includes cleverness which assigns a date-based
version code if a build is done from a raw CVS checkout.

The library now includes a new function, ttodatav(), like ttodata() but
in some cases it gives more verbose error reports.  Also, there are now
a couple of functions which distill an RSA public key to a "key ID" for
use in messages, a crypto-quality pseudo-random-number generator, and a
function that converts a single address into a (singleton) subnet.

Barf's secret/key-censoring utilities have been tidied up so that what
they emit matches the key-ID library functions. 

The ipsec command now supplies a couple more environment variables to
things it calls, and this has reduced the number of things which need to
be edited at build time to know about configuration variables. 

The configuration reader now allows parameter names to begin with "_";
these are strictly reserved for internal development purposes and will
not in general be documented.


As usual, there are assorted small bug fixes and improvements to docs
and messages.



New and notable in 1.96:

Compatibility with the 2.0.xx kernels (under Red Hat 5.2) is broken.
This was accidental, and we have some hopes of fixing it, although it
won't last forever even if we do.  See README.

Our "IKE Implementation Issues" IETF draft has been added to our docs.

Newhostkey now has a mandatory --output "option" specifying a filename
(although the magic filename "-" means "standard output"), and sets
restrictive permissions on newly-created files to try to keep private keys
truly private. 

Pluto now gracefully handles a couple of odd error cases (missing private
keys, network errors at just the wrong time in negotiations) which used to
cause it to die, provoking an automatic restart of the IPsec subsystem. 

Makefile.inc now has more detailed control over where manpages go, to ease
integration into distributions.  The name of the destination directory for
RPMs is also now centralized there.

Rsasigkey now checks for certain obscure internal failures, which can
happen if --random is used to supply a source of random bits that aren't
really very random, and complains accordingly.

Barf's eroute table, formerly sorted by source, is now sorted by destination.
This is probably a bit clearer for hosts with lots of eroutes.

Use of the %reject target in a shunt eroute now, as long planned, actually
causes an ICMP rejection message (Destination Unreachable:  Communication
Administratively Prohibited). 

KLIPS now declares its module license to be "Dual BSD/GPL".  (It is 
dual because the libdes and zlib code is BSD licensed, while KLIPS itself 
is GPL).  Whether this is exactly right is not yet clear, but at least it
stops the latest module loaders from whining.

The UML test stuff continues to grow and add new test cases.

As usual, there are assorted small bug fixes and improvements to docs
and messages.



New and notable in 1.95:

The last remnants of the "%hold" bug, which broke 1.93 and 1.94, have
(we think) been dealt with.

Minor bugs in the RPM-building stuff have been fixed, and it has been
updated to deal with Red Hat's abrupt removal of the --buildroot option
from rpm.

Pluto now supports the ability to delete an instance of a conn, rather
than just the whole conn (for cases, notably OE, where a single connection
description can be involved in multiple connections).  There is no support
for this in ipsec_auto yet.

Pluto's routine (non-debug) logging is now somewhat terser, and some of
the more boring messages are produced only if --debug-lifecycle is turned
on (this is a temporary hack, it's not really debugging output).

Pluto's innards have been reorganized somewhat, in preparation for making
DNS lookups asynchronous.  There should be no user-visible effects.

Pluto now ignores IPv6 link-local interface addresses; we think that they
are never relevant, and trying to bind to them causes difficulties.  This

user/pam_krb5/src/afs5log.c
/*
 * Copyright 2004,2005,2006,2007,2008 Red Hat, Inc.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, and the entire permission notice in its entirety,
 *    including the disclaimer of warranties.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. The name of the author may not be used to endorse or promote
 *    products derived from this software without specific prior
 *    written permission.
 *
 * ALTERNATIVELY, this product may be distributed under the terms of the
 * GNU Lesser General Public License, in which case the provisions of the
 * LGPL are required INSTEAD OF the above restrictions.
 *
 * THIS SOFTWARE IS PROVIDED ``AS IS'' AND ANY EXPRESS OR IMPLIED
 * WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.  IN
 * NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
 * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
 * NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF
 * USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
 * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
 * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#include "../config.h"

#include <sys/types.h>
#include <limits.h>
#include <stdarg.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#include KRB5_H
#ifdef USE_KRB4
#include KRB4_DES_H
#include KRB4_KRB_H
#ifdef KRB4_KRB_ERR_H
#include KRB4_KRB_ERR_H
#endif
#endif
#include <security/pam_appl.h>

#include "init.h"
#include "logstdio.h"
#include "options.h"
#include "stash.h"
#include "minikafs.h"
#include "xstr.h"

extern char *log_progname;

static void
set_methods(const char *strategy, int *methods, int max_methods,
	    int try_v5_2b_only, int try_rxk5_only)
{
	int i = 0, j;
	struct {
		const char *name; int method;
	} method_names[] = {
#ifdef USE_KRB4
		{"v4", MINIKAFS_METHOD_V4},
		{"524", MINIKAFS_METHOD_V5_V4},
#endif
		{"2b", MINIKAFS_METHOD_V5_2B},
		{"rxk5", MINIKAFS_METHOD_RXK5}
	};
	const char *p, *q;

	memset(methods, 0, sizeof(methods[0]) * max_methods);
	if (try_rxk5_only) {
		methods[i++] = MINIKAFS_METHOD_RXK5;
	} else {
		if (try_v5_2b_only) {
			methods[i++] = MINIKAFS_METHOD_V5_2B;
		} else {
			p = strategy;
			while (strlen(p) > 0) {
				q = p + strcspn(p, ",");
				for (j = 0;
				     j < sizeof(method_names) /
				         sizeof(method_names[0]);
				     j++) {
					if (strncmp(p,
						    method_names[j].name,
						    strlen(method_names[j].name)) == 0) {
						methods[i++] = method_names[j].method;
					}
				}
				p = q + strspn(q, ",");
			}
		}
	}
}

int
main(int argc, char **argv)
{
	char local[PATH_MAX], home[PATH_MAX], path[PATH_MAX];
	char *homedir, *cell, *principal, *pathdir, *strategy;
	int i, j, try_v5_2b_only, try_rxk5_only, cells, process_options;
	int methods[8];
	krb5_context ctx;
	krb5_ccache ccache;
	uid_t uid;

	/* Iterate through every parameter, assuming they're names of cells. */
#ifdef USE_KRB4
	try_v5_2b_only = 0;
#else
	try_v5_2b_only = 1;
#endif
	try_rxk5_only = 0;
	cells = 0;
	log_progname = "afs5log";
	strategy = DEFAULT_TOKEN_STRATEGY;
	uid = getuid();
	memset(&log_options, 0, sizeof(log_options));
	memset(&ccache, 0, sizeof(ccache));
	i = _pam_krb5_init_ctx(&ctx, 0, NULL);
	if (i != 0) {
		fprintf(stderr, "Error initializing Kerberos: %d\n", i);
		exit(1);
	}
	i = krb5_cc_default(ctx, &ccache);
	if (i != 0) {
		fprintf(stderr, "Error reading default credential cache: %d\n",
			i);
		exit(1);
	}
	process_options = 1;
	for (i = 1; i < argc; i++) {
		if (process_options && (argv[i][0] == '-')) {
			switch (argv[i][1]) {
			case '-':
				process_options = 0;
				break;
			case 'v':
				log_options.debug++;
				break;
			default:
				break;
			}
		}
	}
	if (minikafs_has_afs() == 0) {
		if (log_options.debug) {
			debug("afs support not detected\n");
		}
		return -1;
	}
	process_options = 1;
	for (i = 1; i < argc; i++) {
		if (process_options && (argv[i][0] == '-')) {
			switch (argv[i][1]) {
			case '-':
				process_options = 0;
				break;
			case '5':
				try_v5_2b_only = !try_v5_2b_only;
				break;
			case 'k':
				try_rxk5_only = !try_rxk5_only;
				break;
			case 'n':
				log_options.null_afs_first =
					!log_options.null_afs_first;
				break;
			case 'v':
				break;
			case 'p':
				i++;
				pathdir = argv[i];
				j = minikafs_cell_of_file_walk_up(pathdir, path,
								  sizeof(path));
				if ((j == 0) &&
				    (strcmp(path, "dynroot") != 0)) {
					cells++;
					if (log_options.debug) {
						debug("cell of \"%s\" is "
						      "\"%s\"", pathdir, path);
					}
					set_methods(strategy, methods,
						    sizeof(methods) /
						    sizeof(methods[0]),
						    try_v5_2b_only,
						    try_rxk5_only);
					j = minikafs_log(NULL, ccache,
							 &log_options,
							 path, NULL,
							 uid, methods, -1);
					if (j != 0) {
						fprintf(stderr,
							"%s: %d\n", path, j);
					}
				}
				break;
			case 's':
				i++;
				strategy = argv[i];
				break;
			default:
				printf("%s: [ [-v] [-5] [-k] [-n] "
				       "[-s strategy] [-p path] "
				       "[cell[=principal]] ] [...]\n",
				       argv[0]);
				krb5_free_context(ctx);
				exit(0);
				break;
			}
		} else {
			cells++;
			cell = xstrdup(argv[i]);
			principal = strchr(cell, '=');
			if (principal != NULL) {
				*principal = '\0';
				principal++;
			}
			set_methods(strategy, methods,
				    sizeof(methods) / sizeof(methods[0]),
				    try_v5_2b_only, try_rxk5_only);
			j = minikafs_log(NULL, ccache, &log_options,
					 cell, principal, uid, methods, -1);
			if (j != 0) {
				fprintf(stderr, "%s: %d\n", argv[i], j);
			}
			xstrfree(cell);
		}
	}

	/* If no parameters were offered, go for the user's home directory and
	 * the local cell, if we can determine what its name is. */
	if (cells == 0) {
		j = minikafs_ws_cell(local, sizeof(local));
		if ((j == 0) && (strcmp(local, "dynroot") != 0)) {
			if (log_options.debug) {
				debug("local cell is \"%s\"", local);
			}
			set_methods(strategy, methods,
				    sizeof(methods) / sizeof(methods[0]),
				    try_v5_2b_only, try_rxk5_only);
			j = minikafs_log(NULL, ccache, &log_options,
					 local, NULL, uid, methods, -1);
			if (j != 0) {
				fprintf(stderr, "%s: %d\n", local, j);
			}
		}
		if (getenv("HOME") != NULL) {
			homedir = xstrdup(getenv("HOME"));
		} else {
			homedir = xstrdup("/afs");
		}
		if (homedir != NULL) {
			j = minikafs_cell_of_file_walk_up(homedir, home,
							  sizeof(home));
			if ((j == 0) &&
			    (strcmp(home, "dynroot") != 0) &&
			    (strcmp(home, local) != 0)) {
				if (log_options.debug) {
					debug("home cell is \"%s\"", home);
				}
				set_methods(strategy, methods,
					    sizeof(methods) /
					    sizeof(methods[0]),
					    try_v5_2b_only, try_rxk5_only);
				j = minikafs_log(NULL, ccache,
						 &log_options,
						 home, NULL, uid,
						 methods, -1);
				if (j != 0) {
					fprintf(stderr, "%s: %d\n", home, j);
				}
			}
			xstrfree(homedir);
		}
	}
	krb5_cc_close(ctx, ccache);
	krb5_free_context(ctx);
	return 0;
}
